### PR TITLE
Feature/tsp solver

### LINKED ISSUE #4: Implement a base TSP solver class
Implement a base class for all TSP ([Travelling Salesman Problem](https://en.wikipedia.org/wiki/Travelling_salesman_problem)) solvers to use on our previously built distance matrix.
The base class should have an abstract "solve" method that takes a distance matrix and outputs a list of nodes indices that forms an [hamiltonian cycle](https://en.wikipedia.org/wiki/Hamiltonian_path).

### LINKED ISSUE #7: Add tests for the TSP solvers
Add a test file in a tests folder to check the TSP solvers on a dataset.

### LINKED ISSUE #11: Implement a TSP solver
Implement a TSP solver class based on the base class. Either reimplement an existing algorithm or use an existing library.

### LINKED ISSUE #36: The built graph can be disconnected
In Nantes data, there are two stops not linked to any other part of the graph : 

Image

This produces inf values in the distance_matrix, which prevent the solver from working. This should be handled accordingly.

### PR DESCRIPTION
**Pull request by KarlLbs**
Implement a first TSP (Travelling Salesman Problem) solver using Google's library OR-tools.
- Solver base class
- A first implementation as ORToolsSolver
- A test using Google's [example](https://developers.google.com/optimization/routing/tsp)
Closes #4 #7 #11 
Waiting for #23 to merge to test on the real data.

A few architecture questions : 
- Should I keep the code I put in the init of the class here, or should I move it to the solve method ? 
- If we want to solve for different starting points, we will need to change the initialization of the index_manager, that, for now, is in the init. So I need to add this parameter as an argument, but I feel like it's probably better as an argument of the solve method. If it is the case, then this solves my previous question since initializing the index_manager is the first line of the init, so I need to move everything to the solve method.
- Similarly, should the distance matrix be argument of the init or the solve method ? Maybe distance matrix as an argument to the init and the starting point as an argument to the solve ?

@Polemique

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -16,6 +16,7 @@ dependencies = [
     "pyproj",
     "shapely",
     "tqdm",
+    "ortools>=9.15.6755",
```
Comment (Polemique) [inline pyproject.toml]: 👀

File: src/hamilbus/solvers/base_solver.py
Comment (Polemique) [inline src/hamilbus/solvers/base_solver.py:1]: Great habit

File: src/hamilbus/solvers/or_tools_solver.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
+### or_tools_solver.py
+### Implement a simple TSP solver using Google's OR-Tools
+
+from hamilbus.solvers.base_solver import BaseSolver
```
Comment (Polemique) [inline src/hamilbus/solvers/or_tools_solver.py:6]: You don't want hbus right ? 👀

File: src/hamilbus/solvers/or_tools_solver.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+from ortools.constraint_solver import routing_enums_pb2
+from ortools.constraint_solver import pywrapcp
```
Comment (Polemique) [inline src/hamilbus/solvers/or_tools_solver.py]: One ligne of code ?

File: src/hamilbus/solvers/or_tools_solver.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+
+class ORToolsSolver(BaseSolver):
```
Comment (Polemique) [inline src/hamilbus/solvers/or_tools_solver.py:10]: Could we have docstring about the method ?

**Comment by KarlLbs** [comment]
New and improved version, and I answered the questions I was asking at the beginning.
It now runs on the real data and produced a solution. Closes #36 

Please review @Polemique 🙏

File: src/hamilbus/helpers.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+def save_array(arr: np.ndarray, path: str | Path) -> Path:
+    """Save a NumPy array to disk. Extension is set automatically (.npy) if omitted"""
+    path = Path(path).with_suffix(".npy")
+    if not path.is_absolute():
+        path = Path.cwd() / path
+    np.save(path, arr)
+    return path.resolve()
+
+
+def load_array(path: str | Path) -> np.ndarray:
```
Comment (Polemique) [inline src/hamilbus/helpers.py]: This is not what np.load() already do ?

File: src/hamilbus/helpers.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+def save_dict(d: dict[float, str], path: str | Path) -> Path:
+    """Saves a dict as a json file in the current working dir"""
+    path = Path.cwd() / Path(path).with_suffix(".json")
+    with open(path, "w", encoding="utf-8") as f:
+        json.dump({str(k): v for k, v in d.items()}, f, indent=2)
+    return path.resolve()
+
+
+def load_dict(path: str | Path) -> dict[float, str]:
```
Comment (Polemique) [inline src/hamilbus/helpers.py]: Same comment as load array

File: src/hamilbus/helpers.py
Comment (Polemique) [inline src/hamilbus/helpers.py:1]: There is a lot of time the same verifications (path exist ?), could we refactor DRY ?

File: src/hamilbus/main.py [hunk 7]
```diff
@@ -1,18 +1,35 @@
...
+from hamilbus.solvers.or_tools_solver import ORToolsSolver
+from hamilbus.distance_matrix import compute_distance_matrix
 
 
 def main() -> None:
     """Start the Hamilbus local web viewer on port 3000."""
+    # Check if there's already a distance_matrix saved
+    path = Path.cwd() / "distance_matrix.npy"
+    if path.exists():
+        distance_matrix = load_array("distance_matrix.npy")
```
Comment (Polemique) [inline src/hamilbus/main.py]: Load array verify if path exist, why double protection?

File: src/hamilbus/main.py [hunk 8]
```diff
@@ -1,18 +1,35 @@
...
+    else:
+        # Load the raw data
+        DATA_DIR = Path(__file__).resolve().parents[1] / "hamilbus" / "data"
+        stops, lines, trips_by_lines, stops_by_trips = reader.load_gtfs(DATA_DIR)
+        # Build the graph
+        graph_builder = GraphBuilder(stops, lines, trips_by_lines, stops_by_trips)
+        graph_builder.merge_stops()
+        graph = graph_builder.build_graph()
```
Comment (Polemique) [inline src/hamilbus/main.py:24]: I like how the builder is use

File: src/hamilbus/solvers/or_tools_solver.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+# faire un truc pour mettre la node de départ comme input avec un nom/id d'arret
+
+
+class ORToolsSolver(BaseSolver):
+    def __init__(self, distance_matrix: np.ndarray, time_limit_seconds: int = 30):
+        self.distance_matrix = self._transform_distance_matrix(distance_matrix)
+        self.time_limit_seconds = time_limit_seconds
+
+    def _transform_distance_matrix(self, distance_matrix):
```
Comment (Polemique) [inline src/hamilbus/solvers/or_tools_solver.py]: "Transform" is very general, transform in what ?

File: src/hamilbus/solvers/or_tools_solver.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            path.append(self.index_manager.IndexToNode(index))
+            previous_index = index
+            index = solution.Value(self.routing_model.NextVar(index))
+            route_distance += self.routing_model.GetArcCostForVehicle(
+                previous_index, index, 0
+            )
+        path.append(self.index_manager.IndexToNode(index))
+        return path, route_distance
+
+    def print_solution(self, solution):
```
Comment (Polemique) [inline src/hamilbus/solvers/or_tools_solver.py:107]: I like using print method

File: tests/test_helpers.py
Comment (Polemique) [inline tests/test_helpers.py:1]: Ok

File: tests/test_solvers.py
Comment (Polemique) [inline tests/test_solvers.py:1]: Could we have a test with no solution ?

**Review by Polemique — APPROVED** [review]
Great, pending comment but +2 after handling it. 

And let's go try to DRY 😎

File: pyproject.toml (hunk 1 above)
Comment (KarlLbs) [reply pyproject.toml]: Changed in 6262d102e95dae0a4a428820a7a4e5fdfa54e54e

File: src/hamilbus/solvers/or_tools_solver.py (hunk 3 above)
Comment (KarlLbs) [reply src/hamilbus/solvers/or_tools_solver.py]: Done in 9831b9e34098032593f0475cc45edde70709d1ea

File: src/hamilbus/solvers/or_tools_solver.py (hunk 4 above)
Comment (KarlLbs) [reply src/hamilbus/solvers/or_tools_solver.py:10]: Added a class docstring in 9831b9e34098032593f0475cc45edde70709d1ea

File: src/hamilbus/solvers/or_tools_solver.py (hunk 9 above)
Comment (KarlLbs) [reply src/hamilbus/solvers/or_tools_solver.py]: Renamed to preprocess in 9831b9e34098032593f0475cc45edde70709d1ea

File: tests/test_solvers.py
Comment (KarlLbs) [reply tests/test_solvers.py:1]: Multiple edge cases tests added in 5140b7362cd2b2b6e51d49dc97fa39a790c2ed15

File: src/hamilbus/helpers.py (hunk 5 above)
Comment (KarlLbs) [reply src/hamilbus/helpers.py]: Helpers deleted in 16b32a4e9943f2d7e97d241ee34625deeabd2ef7

File: src/hamilbus/helpers.py (hunk 6 above)
Comment (KarlLbs) [reply src/hamilbus/helpers.py]: Same 16b32a4e9943f2d7e97d241ee34625deeabd2ef7

File: src/hamilbus/helpers.py
Comment (KarlLbs) [reply src/hamilbus/helpers.py:1]: Helpers deleted 16b32a4e9943f2d7e97d241ee34625deeabd2ef7

File: src/hamilbus/main.py (hunk 7 above)
Comment (KarlLbs) [reply src/hamilbus/main.py]: It's not for protection here but for scenario handling : if there's no matrix saved, we compute one
